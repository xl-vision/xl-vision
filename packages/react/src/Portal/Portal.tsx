import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import React from 'react';
import { isProduction, isServer } from '@xl-vision/utils';
import getContainer, { ContainerReturnType, ContainerType } from '../utils/getContainer';

export type {
  ContainerType as PortalContainerType,
  ContainerReturnType as PortalContainerReturnType,
};

export interface PortalProp {
  children?: React.ReactNode;
  container?: ContainerType;
}

/**
 * 支持服务端渲染
 * getContainer支持传递父元素，字符串或者undefined,也可以通过函数返回
 * 如果传递字符串，则根据字符串进行查询节点
 * 如果是undefined，则直接挂载在当前位置
 * @param props
 * @constructor
 */
const Portal: React.FC<PortalProp> = (props) => {
  const { children, container: containerProp } = props;

  if (isServer) {
    return null;
  }

  const container = getContainer(containerProp);

  if (!container) {
    return <>{children}</>;
  }

  return ReactDOM.createPortal(children, container);
};

if (!isProduction) {
  Portal.displayName = 'Portal';

  Portal.propTypes = {
    children: PropTypes.element.isRequired,
    container: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.string,
      isServer ? PropTypes.any : PropTypes.instanceOf(Element),
    ]),
  };
}

export default Portal;
