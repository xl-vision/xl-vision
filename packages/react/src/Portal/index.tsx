import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import React from 'react';
import { isServer } from '../utils/env';
import warning from '../utils/warning';

export type PortalContainerReturnType = HTMLElement | string | undefined | null;
export type PortalContainerType = PortalContainerReturnType | (() => PortalContainerReturnType);

export interface PortalProp {
  children: React.ReactElement;
  getContainer?: PortalContainerType;
}

/**
 * 支持服务端渲染
 * getContainer支持传递父元素，字符串或者undefined,也可以通过函数返回
 * 如果传递字符串，则根据字符串进行查询节点
 * 如果是undefined，则直接挂载在当前位置
 * @param props
 * @constructor
 */
const Portal: React.FunctionComponent<PortalProp> = (props) => {
  const { children, getContainer } = props;
  if (isServer) {
    return null;
  }

  let container: PortalContainerType;

  if (typeof getContainer === 'function') {
    container = getContainer();
  } else {
    container = getContainer;
  }

  if (!container) {
    return children;
  }

  if (typeof container === 'string') {
    const selector = document.querySelector(container);
    warning(!selector, '<Portal> querySelector "%s" is null', container);
    return ReactDOM.createPortal(children, selector!);
  }

  return ReactDOM.createPortal(children, container);
};

Portal.displayName = 'Portal';

Portal.propTypes = {
  children: PropTypes.element.isRequired,
  getContainer: PropTypes.any,
};

export default Portal;
