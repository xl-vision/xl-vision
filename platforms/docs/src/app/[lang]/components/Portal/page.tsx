import { componentRoutes } from '@docs/routes';
import createGenerateMetadata from '@docs/utils/createGenerateMetadata';
import Docs from './_docs';

export const generateMetadata = createGenerateMetadata(componentRoutes, 'Portal');

export default Docs;
