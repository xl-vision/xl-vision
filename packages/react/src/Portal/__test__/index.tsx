import { mount } from 'enzyme';
import React from 'react';
import Portal from '..';

describe('Portal', () => {
  it('test mount on body', () => {
    mount(
      <div>
        <Portal getContainer={() => document.body}>
          <div>123</div>
        </Portal>
      </div>,
    );

    expect(document.body).toMatchSnapshot();
  });

  it('test mount on current position', () => {
    const wrapper = mount(
      <div>
        <Portal getContainer={null}>
          <div>123</div>
        </Portal>
      </div>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });
});
