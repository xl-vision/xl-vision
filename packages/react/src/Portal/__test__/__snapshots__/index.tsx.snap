// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal test mount on body 1`] = `
<body>
  <div>
    123
  </div>
</body>
`;

exports[`Portal test mount on current position 1`] = `
<div>
  <div>
    123
  </div>
</div>
`;
